# Scripts Documentation fo Xarxa

All scripts used to interact with the database are located in the `src` directory,
although some pre-build workflows are provided in the `workflows` directory.

All scripts provide a help message that can be accessed by running the script with the `-h` or `--help` option. The help message will provide information about the script's purpose, usage, options and some examples.

All scripts print expected output (if any) to the standard output. Any
other information, such as logs, warnings, and error messages, are printed to the standard error.

The scripts are designed to be run in pipelines, where the output of one script is the input of the next one. Data fetching and processing has been separated for easier debugging and maintenance.

Data format of choice is a tabular file, where each row represents a record and each column represents a field for that record. Columns are separated by tabs. Columns may contain different types of data, such as strings, integers, arrays, or no value at all. Each type
representation is described below:
- Strings are represented as they are.
- Integers are represented as they are.
- Arrays are represented as a string with the elements separated by semicolons.
- Dictionary and JSON objects are represented as canonic JSON strings.
- The NoneType value is represented as a "NULL" string.

*Example of and hypotetical tabular file generated by the scripts:*

```text
P12345  LT1;LT2;LT3  R12345  +  1  100  MAAKQ
P12346  LT4  R12346  -  200  300  MAAKQ
P12347  NULL R12347  +  400  500  MAAKQ
P12348  LT5  R12348;R12349  -  600  700  NULL
```

## Table of Contents

- [Script Listings](#script-listings)
    - [Data Fetching Scripts](#data-fetching-scripts)
    - [Data Preprocessing Scripts](#data-preprocessing-scripts)
    - [Data Insertion Scripts](#data-insertion/Update-scripts)
    - [Utility Scripts](#utility-scripts)
- [Script Details](#script-details)
    - [create_id_master_table.py](#create_id_master_tablepy)
    - [fetch_uniprot_organism_ids.py](#fetch_uniprot_organism_idspy)
    - [upsert_table.py](#upsert_tablepy)
        - [uniprot_organism](#uniprot_organism)
        - [refseq_genome](#refseq_genome)

## Script Listings

### Data Fetching Scripts

Scripts used to fetch data from external sources. These scripts are designed to retrieve data from various sources and format it for insertion into the database.

The fetched data is printed to stdout in a tab-separated format. **No headers are included in the output.**

#### Links:

- [`fetch_kegg_organism.py`](#fetch_kegg_organismpy)
- [`fetch_kegg_relations.py`](#fetch_kegg_relationspy)
- [`fetch_refseq_genome.py`](#fetch_refseq_genomepy)
- [`fetch_uniprot_organism_json_entry.py`](#fetch_uniprot_organism_json_entrypy)

### Data Preprocessing Scripts

Scripts that preprocess data before insertion into the database.

The formatted data is printed to stdout in a tab-separated format. **No headers are included in the output.**

#### Links:

- [`process_uniprot_json_entry.py`](#process_uniprot_json_entrypy)

### Data Insertion/Update Scripts

Scripts that facilitate the insertion of new data into the database. This includes protocols for handling and validating new entries.

#### Links:

- [`upsert_table.py`](#upsert_tablepy)

### Utility Scripts

Scripts designed to update existing records within the database. These scripts ensure data remains current and correct.
Maintenance Scripts

Scripts for database maintenance tasks such as backup, data cleaning, and performance optimization.

#### Links:


- [`printval.py`](#printvalpy)
- [`tab_to_fasta.py`](#tab_to_fastapy)

## Script Details

### fetch_kegg_organism.py

#### Description

Fetches all KEGG entries for a specified organism.

#### Usage

```shell
python src/fetch_kegg_organism.py [options] <kegg_organism_id>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<kegg_organism>` | KEGG organism ID | String | - | Yes |
| `--log` | log file path | string | info | no |
| `-h`, `--help` | show help message | - | - | no |

#### Output

| Column name | Type | Description |
|-------------|------|-------------|
| KEGG Accession | String | The KEGG accession number |
| Pathways | Array\[String\] | The pathways associated with the entry, separated by semicolons |
| KEGG Orthology | Array\[String\] | The KEGG Orthology (KO) numbers associated with the entry, separated by semicolons |

#### Examples

Fetch KEGG IDs for a specific organism:
```bash
python src/fetch_kegg_organism.py sml
```

Save fetched data to a file:
```bash
python src/fetch_kegg_organism.py sml > kegg_data.txt
```

Chain data fetching and processing:
```bash
python src/fetch_kegg_organism.py sml | tee data.log | python process_data.py
```

### fetch_kegg_relations.py

#### Description

Given a KEGG organism code, fetches all the relations for each pathway in the organism.

#### Usage

```shell
python src/fetch_kegg_relations.py [options] <kegg_organism_id>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<kegg_organism>` | KEGG organism ID | String | - | Yes |
| `--log` | log file path | string | info | no |
| `-h`, `--help` | show help message | - | - | no |

#### Output

| Column name | Type | Description |
|-------------|------|-------------|
| Source KEGG Accession | String | The source KEGG accession number |
| Target KEGG Accession | String | The target KEGG accession number |
| Pathway ID | String | The pathway ID |
| Relation Type | String | The relation type |
| Relation Subtypes | Array\[String\] | The relation subtypes, separated by semicolons |
| Relation Subtype Values | Array\[String\] | The relation subtype values, separated by semicolons. Mostly used when referring to a compound name |

#### Examples

Fetch relations for a specific organism:
```bash
python src/fetch_kegg_relations.py sml
```

Chain data fetching and processing:
```bash
python src/fetch_kegg_relations.py sml | tee data.log | python process_data.py
```

### fetch_uniprot_organism_json_entry.py

#### Description

Fetches all UniProt entries for a specified organism. The script fetches the JSON entry for each UniProt entry and prints it to stdout.


#### Usage

```shell
python src/fetch_uniprot_organism_json_entry.py [options] <taxon_id>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<taxon_id>` | Taxon ID of the organism | Integer | - | Yes |
| `--log` | log file path | string | info | no |
| `-h`, `--help` | show help message | - | - | no |

#### Output

A JSON file containing `results` as the root key. Each entry in the `results` array is a JSON object representing a UniProt entry.


#### Examples

Fetch UniProt entries for a specific taxon ID:
```bash
python src/fetch_uniprot_organism_json_entry.py 522373
```

EOF
----- 

### fetch_refseq_genome.py

#### Description

Extracts data from a RefSeq annotated genome in GenBank format. Formats extracted data and prints it in tabular format to stdout for downstream processing.

#### Usage

```shell
python src/fetch_refseq_genbank_genome.py [options] <file>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | GenBank file path, or '-' for stdin | String | '-' | Yes |
| `--log` | log file path | string | info | no |
| `-h`, `--help` | show help message | - | - | no |

#### Output

| Column name | Type | Description |
|-------------|------|-------------|
| RefSeq Locus Tag | String | The RefSeq locus tag |
| Locus Tag | Array\[String\] | The locus tag(s) of the protein, separated by semicolons |
| RefSeq Accession | String | The RefSeq accession number |
| Strand Location | String | The strand location of the protein |
| Start Position | String | The start position of the protein |
| End Position | String | The end position
| Translated Protein Sequence | String | The translated protein sequence |

#### Examples

Fetch data from a GenBank file:
```bash
python src/fetch_refseq_genbank_genome.py data/genbank_file.gb
```

Chain data fetching and processing:
```bash
python src/fetch_refseq_genbank_genome.py data/genbank_file.gb | tee data.log | python process_data.py
```

Obtain context for a specific gene:
```bash
python src/fetch_refseq_genbank_genome.py data/genbank_file.gb | grep -C 10 "MyFavoriteLocusTag"
```

#### Troubleshooting

If a feature is a pseudogene, the build-in functions from Biopython are used to
translate the sequence. When translating a pseudogene,
the translation table is not specified and when a stop codon is found, the
translation is stopped.

### fetch_uniprot_organism_ids.py

#### Description

Retrieves all UniProt database entries for a specified organism given its taxon ID.
Collects relevant IDs and mappings for each entry and prints them to stdout in a tab-separated format.

#### Usage

```shell
python src/fetch_uniprot_organism_ids.py [options] <taxon_id>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<taxon_id>` | Taxon ID of the organism | Integer | - | Yes |
| `--log` | log file path | string | info | no |
| `-h`, `--help` | show help message | - | - | no |

#### Output

A tab-separated file including:

| Column name | Type | Description |
|-------------|------|-------------|
| UniProt Primary Accession | String | The UniProtKB primary accession number |
| Locus Tag | Array\[String\] | The locus tag(s) of the protein, separated by semicolons |
| RefSeq Accession | Array\[String\] | The RefSeq accession number(s), separated by semicolons |
| EMBL Protein ID | String | The EMBL protein ID |
| KEGG Accession | Array\[String\] | The KEGG accession number(s), separated by semicolons |

#### Examples

Fetch UniProt IDs for a specific taxon ID:
```bash
python src/fetch_uniprot_organism_ids.py 522373
```

Save fetched data to a file:
```bash
python src/fetch_uniprot_organism_ids.py 522373 > uniprot_data.txt
```

Chain data fetching and processing:
```bash
python src/fetch_uniprot_organism_ids.py 522373 | tee data.log | python process_data.py
```

#### Troubleshooting

Example of API request made by the script: <https://rest.uniprot.org/uniprotkb/stream?query=organism_id:522373&fields=accession,gene_oln,gene_orf,xref_refseq,xref_kegg&format=tsv>

Notice that the fields requested the UniProt API are actually:
- `accession`: UniProt Primary Accession
- `gene_oln`: Ordered Locus Name. See UniProt [documentation](<https://www.uniprot.org/help/gene_name>) for more information.
- `gene_orf`: Open Reading Frame. See UniProt [documentation](<https://www.uniprot.org/help/gene_name>) for more information.
- `xref_refseq`: RefSeq Accession
- `xref_kegg`: KEGG Accession

Both `xref_refseq` and `xref_kegg` are arrays, so they will be represented as 
strings with the elements separated by semicolons. If not empty, an additional
semicolon is always added at the end of the string.

Surprisingly, while `gene_oln` and `gene_orf` are also arrays, an empty space
character is used as a separator. I  wouldn't be surprised if at some point they
decide to change the separator character. **After retreiving the data, both fields will
be merged into the Old Locus Tag column**

The EMBL Protein ID is not an available field in the UniProt API. The solution
is to access the JSON file for each entry individually and extract the EMBL Protein ID from it.

### generate_enums.py

#### Description

Connect to the database and generate/update the enum types based on the configuration file.

#### Usage

```shell
python src/generate_enums.py [options]
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Output

None

### preprocess_peptide_ptm_data.py

#### Description

Preprocess Proteomics Experimental Data.

More specifically, this script preprocesses the peptide PTM data produced by
the Thermo Fisher Scientific Proteome Discoverer software (version 2.5).
Accepts a single input file in tab-delimited format, containing **ONLY** the
following columns:
    1. Protein ID (Master Protein Accessions)
    2. PTMs (Modifications)
    3. Peptide Sequence (Sequence)
    4. Position in Master Protein (Position in Master Protein)
    5. PSM Ambiguity (PSM Ambiguity)
    6. Posterior Error Probability (PEP)
    7. q-value (q-value)
    8. Number of Protein Groups (# Protein Groups)
    9. Search Engine Confidence (Confidence)

Essentially, the script will retrain any record that has PTMs in the "PTMs"
column and generate one record for every PTM found in every peptide sequence.

#### Usage

```shell
python src/preprocess_peptide_ptm_data.py [options] <file>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |


#### Input Format

The expected input format is a tab-separated file. The `.xlsx` file output by
Proteome Discoverer can be converted to a tab-separated file using Excel. Then, 
a single file containing the columns mentioned below is expected.

| Column name | Type | Description |
|-------------|------|-------------|
| Protein ID | String | The ID of the protein |
| PTMs | String | The modifications |
| Peptide Sequence | String | The sequence of the peptide |
| Position in Master Protein | Integer | The position in the master protein |
| PSM Ambiguity | Integer | The PSM ambiguity |
| Posterior Error Probability | Float | The posterior error probability |
| q-value | Float | The q-value |
| Number of Protein Groups | Integer | The number of protein groups |
| Search Engine Confidence | String | The confidence of the search engine |

#### Output


### printval.py

#### Description

Print tables, columns, values or types from the database.

#### Usage

```shell
python src/printval.py [options] <table> <columns>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<table>` | Name of the database table to query. Use 'list' to display all tables. Use 'enum' to display the enum types in the database. | String | 'list' | Yes |
| `<columns>` | Specific columns to display. Use commas for separating multiple columns. Use 'list' to display all columns. Use the enum type name to display all values of the enum type if previous argument is 'enum'. | String | 'list' | No |
| `--protein-type` | If the table contains a 'protein_type' column, filter by this value. See 'protein_type' enum in the database schema for possible values. | String | - | No |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Output

A tabular file with the requested data. Format follows same conventions as the rest of the scripts.

#### Examples

Display all tables in the database:
```shell
python src/printval.py list
```

Display all columns from the 'uniprot_protein' table:
```shell
python src/printval.py uniprot_protein list
```

Fetch all 'uniprot_accession' from 'uniprot_protein':
```shell
python src/printval.py uniprot_protein uniprot_accession
```

Retrieve 'uniprot_accession', 'go_function', and 'go_process' from 'uniprot_protein':
```shell
python src/printval.py uniprot_protein uniprot_accession,go_function,go_process
```

Retrieve 'uniprot_accession' from 'uniprot_protein' where 'protein_type' is 'kinase':
```shell
python src/printval.py uniprot_protein uniprot_accession --protein-type kinase
```

Retrive 'target_with_participant' from 'kegg_relations' where 'protein_type' of the query protein is 'kinase':
```shell
python src/printval.py kegg_relations target_with_participant --protein-type kinase
```

List all enum types in the database:
```shell
python src/printval.py enum
```

List all values of the 'protein_type' enum:
```shell
python src/printval.py enum protein_type
```

## upsert_table.py

### Description

Read a TSV file, validate, and upsert the data into a table in a database. Creates the table if it does not exist

### Usage

```shell
python src/upsert_table.py <table_type> [options]
```

To get more information about the upsert operation on each type of table, use:
```shell
python src/upsert_table.py <table_type> -h
```

### Table Types

### uniprot_organism

#### Usage

```shell
python src/upsert_table.py uniprot_organism [options] <file>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log`| Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

The expected input format is a tab-separated file
as the one generated by the `fetch_uniprot_organism_ids.py` script. See [fetch_uniprot_organism_ids.py](#fetch_uniprot_organism_idspy) for more information.

#### Example

```shell
python src/upsert_table.py uniprot_organism data/uniprot_data.tsv
```

```shell
python src/fetch_uniprot_organism_ids.py 522373 | python src/upsert_table.py uniprot_organism
```

### refseq_genome

#### Usage

```shell
python src/upsert_table.py refseq_genome [options] <file>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

The expected input format is a tab-separated file as the one generated by the `fetch_refseq_genbank_genome.py` script. See [fetch_refseq_genbank_genome.py](#fetch_refseq_genbank_genomepy) for more information.

#### Example

```shell
python src/upsert_table.py refseq_genome data/refseq_data.tsv
```

```shell
python src/fetch_refseq_genbank_genome.py data/genbank_file.gb | python src/upsert_table.py refseq_genome
```

### kegg_organism

#### Usage

```shell
python src/upsert_table.py kegg_organism [options] <file>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

The expected input format is a tab-separated file as the one generated by the `fetch_kegg_organism_ids.py` script. See [fetch_kegg_organism_ids.py](#fetch_kegg_organism_idspy) for more information.

#### Example

```shell
python src/upsert_table.py kegg_organism data/kegg_data.tsv
```

```shell
python src/fetch_kegg_organism_ids.py sml | python src/upsert_table.py kegg_organism
```

### uniprot_protein

#### Usage

```shell
python src/upsert_table.py uniprot_protein [options] <file> <protein_type>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `<protein_type>` | Protein type | String | - | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

The expected input format is a tab-separated file as the one generated by the `fetch_uniprot_query.py` script. See [fetch_uniprot_query.py](#fetch_uniprot_querypy) for more information.

#### Example

```shell
python src/upsert_table.py uniprot_protein data/uniprot_protein_data.tsv "<protein_type>"
```

```shell
python src/fetch_uniprot_query.py --keyword "KW-0418" --taxid 9606 | python src/upsert_table.py uniprot_protein "kinase"
```

### kegg_relations

#### Usage

```shell
python src/upsert_table.py kegg_relations [options] <file> <protein_type>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `<protein_type>` | Protein type | String | - | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

The expected input format is a tab-separated file as the one generated by the `fetch_kegg_relations.py` script. See [fetch_kegg_relations.py](#fetch_kegg_relationspy) for more information.

#### Example

```shell
python src/upsert_table.py kegg_relations data/kegg_relations_data.tsv "<protein_type>"
```

```shell
python src/fetch_kegg_relations.py kegg_data.txt | python src/upsert_table.py kegg_relations "kinase"
```

### string_interactions

#### Usage

```shell
python src/upsert_table.py string_interactions [options] <file> <id_type>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `<id_type>` | ID type | String | - | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

See the [Data Preprocessing](DataPreprocessing.html) section for more information about the expected input format.

#### Example

```shell
python src/upsert_table.py string_interactions data/string_interactions_data_formatted.tsv "refseq_locus_tag"
```

### transcriptomics_quantification

#### Usage

```shell
python src/upsert_table.py transcriptomics_quantification [options] <file> <id_type> <sample>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `<id_type>` | ID type | String | - | Yes |
| `<sample>` | Sample name | String | - | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

| Column name | Type | Description |
|-------------|------|-------------|
| ID | String | The ID of the gene |
| Replicate 1 Read Count | Float | The read count for replicate 1 |
| Replicate 2 Read Count | Float | The read count for replicate 2 |
| Replicate 3 Read Count | Float | The read count for replicate 3 |
| Replicate 1 Normalized Read Count | Float | The normalized read count for replicate 1 |
| Replicate 2 Normalized Read Count | Float | The normalized read count for replicate 2 |
| Replicate 3 Normalized Read Count | Float | The normalized read count for replicate 3 |

See the [Data Preprocessing](DataPreprocessing.html) section for more information about how to format the data.

#### Example

```shell
python src/upsert_table.py transcriptomics_quantification data/transcriptomics_quantification_data_formatted.tsv "refseq_locus_tag" "control"
```

### transcriptomics_differential

#### Usage

```shell
python src/upsert_table.py transcriptomics_differential [options] <file> <id_type> <condition>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `<id_type>` | ID type | String | - | Yes |
| `<condition>` | Condition name | String | - | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

| Column name | Type | Description |
|-------------|------|-------------|
| ID | String | The ID of the gene |
| Log2 Fold Change | Float | The log2 fold change |
| P Value | Float | The p-value |
| Adjusted P Value | Float | The adjusted p-value |

See the [Data Preprocessing](DataPreprocessing.html) section for more information about
how to format the data.

#### Example

```shell
python src/upsert_table.py transcriptomics_differential data/transcriptomics_differential_data_formatted.tsv "refseq_locus_tag" "control"
```

### proteomics_quantification

#### Usage

```shell
python src/upsert_table.py proteomics_quantification [options] <file> <id_type> <sample>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `<id_type>` | ID type | String | - | Yes |
| `<sample>` | Sample name | String | - | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

| Column name | Type | Description |
|-------------|------|-------------|
| Protein ID | String | The ID of the protein |
| Sequence | String | The sequence of the protein |
| Sum Posterior Error Probability | Float | The sum of the posterior error probabilities |
| Combined Q-Value | Float | The combined q-value |
| Replicate 1 Abundance | Float | The abundance for replicate 1 |
| Replicate 2 Abundance | Float | The abundance for replicate 2 |
| Replicate 3 Abundance | Float | The abundance for replicate 3 |
| Replicate 1 Normalized Abundance | Float | The normalized abundance for replicate 1 |
| Replicate 2 Normalized Abundance | Float | The normalized abundance for replicate 2 |
| Replicate 3 Normalized Abundance | Float | The normalized abundance for replicate 3 |
| Replicate 1 Abundance Count | Integer | The abundance count for replicate 1 |
| Replicate 2 Abundance Count | Integer | The abundance count for replicate 2 |
| Replicate 3 Abundance Count | Integer | The abundance count for replicate 3 |

See the [Data Preprocessing](DataPreprocessing.html) section for more information about how to format the data.

#### Example

```shell
python src/upsert_table.py proteomics_quantification data/proteomics_quantification_data_formatted.tsv "uniprot_accession" "control"
```

### proteomics_peptide_ptm

#### Usage

```shell
python src/upsert_table.py proteomics_peptide_ptm [options] <file> <id_type> <sample>
```

#### Parameters

| Parameter | Description | Type | Default Value | Required |
|-----------|-------------|------|---------------|----------|
| `<file>` | File containing tabular data, or '-' for stdin | String | '-' | Yes |
| `<id_type>` | ID type | String | - | Yes |
| `<sample>` | Sample name | String | - | Yes |
| `--db` | Database connection string | String | Reads from `config/database.json` | No |
| `--log` | Log file path | String | INFO | No |
| `-h`, `--help` | Show help message | - | - | No |

#### Input Format

| Column name | Type | Description |
|-------------|------|-------------|
| Protein ID | String | The ID of the protein |
| Modification Type | String | The type of modification |
| Site | String | The site of the modification |
| Sequence | String | The sequence of the peptide |
| Start Position in Complete Protein | Integer | The start position of the peptide in the complete protein |
| End Position in Complete Protein | Integer | The end position of the peptide in the complete protein |
| Posterior Error Probability | Float | The posterior error probability |
| Q-Value | Float | The q-value |
| Number of Protein Groups | Integer | The number of protein groups |
| Search Engine Confidence | Text | The search engine confidence |

See the [Data Preprocessing](DataPreprocessing.html) section for more information about how to format the data.

#### Example

```shell
python src/upsert_table.py proteomics_peptide_ptm data/proteomics_peptide_ptm_data_formatted.tsv "uniprot_accession" "control"
```
